File: Python/099__Recover_Binary_Search_Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):

    def inorder(self, root, val_list, node_list):
        if root:
            self.inorder(root.left, val_list, node_list)
            val_list.append(root.val);
            node_list.append(root)
            self.inorder(root.right, val_list, node_list)

    def recoverTree(self, root):
        """
        :type root: TreeNode
        :rtype: void Do not return anything, modify root in-place instead.
        """
        val_list = []
        node_list = []
        self.inorder(root, val_list, node_list)
        val_list.sort()
        for i,v in enumerate(val_list):
            node_list[i].val = v

File: Java/099__Recover_Binary_Search_Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {

    TreeNode val1 = null, val2 = null;
    TreeNode min = null;

    public void recoverTree(TreeNode root) {
        walk(root);
        int tmp = val1.val;
        val1.val = val2.val;
        val2.val = tmp;
    }

    private void walk(TreeNode curr){
        if (curr == null) return;
        walk(curr.left);
        if (min != null && curr.val < min.val) {
            if (val1 == null) {
                val1 = min;
                val2 = curr;
            } else
                val2 = curr; //override previous val2 since there are only two misplaced elems
        }
        min = curr;
        walk(curr.right);
    }
}
